skactiveml/pool/_badge.py

```diff
@@ -0,0 +1,232 @@
+            If candidates is of shape (n_candidates, n_features), the indexing
+            refers to samples in candidates.
+        utilities : numpy.ndarray of shape (batch_size, n_samples) or
+            numpy.ndarray of shape (batch_size, n_candidates)
+            The utilities of samples before each selected sample of the batch,
+            e.g., `utilities[0]` indicates the utilities used for selecting
+            the first sample (with index `query_indices[0]`) of the batch.
+            Utilities for labeled samples will be set to np.nan.
+            For the case where sample choose uniform random from the set, the
+            utility of all samples will be 1.
+            The utilities represent here the probabilities of samples being
+            chosen.
+            If candidates is None or of shape (n_candidates), the indexing
+            refers to samples in X.
+            If candidates is of shape (n_candidates, n_features), the indexing
+            refers to samples in candidates.
+        """
+        # Validate input parameters
+        X, y, candidates, batch_size, return_utilities = self._validate_data(
+            X, y, candidates, batch_size, return_utilities, reset=True
+        )
+
+        X_cand, mapping = self._transform_candidates(candidates, X, y)
+
+        # Validate classifier type
+        check_type(clf, "clf", SkactivemlClassifier)
+        check_equal_missing_label(clf.missing_label, self.missing_label_)
+        check_scalar(fit_clf, "fit_clf", bool)
+        check_scalar(return_embeddings, "return_embeddings", bool)
+
+        # Fit the classifier
+        if fit_clf:
+            clf = clone(clf).fit(X, y)
+
+        # find the unlabeled dataset
+        if candidates is None:
+            X_unlbld = X_cand
+            unlbld_mapping = mapping
+        elif mapping is not None:
+            unlbld_mapping = unlabeled_indices(
+                y[mapping], missing_label=self.missing_label
+            )
+            X_unlbld = X_cand[unlbld_mapping]
+            unlbld_mapping = mapping[unlbld_mapping]
+        else:
+            X_unlbld = X_cand
+            unlbld_mapping = np.arange(len(X_cand))
+
+        # gradient embedding, aka predict class membership probabilities
+        if return_embeddings:
+            probas, X_unlbld = clf.predict_proba(
+                X_unlbld, return_embeddings=True
+            )
+        else:
+            probas = clf.predict_proba(X_unlbld)
+        p_max = np.max(probas, axis=1).reshape(-1, 1)
+        g_x = (p_max - 1) * X_unlbld
+
+        # init the utilities
+        if mapping is not None:
+            utilities = np.full(
+                shape=(batch_size, X.shape[0]), fill_value=np.nan
+            )
+        else:
+            utilities = np.full(
+                shape=(batch_size, X_cand.shape[0]), fill_value=np.nan
+            )
+
+        # sampling with kmeans++
+        query_indicies = np.array([], dtype=int)
```

Comment:
```suggestion
        query_indicies = []
```
List are faster when appending ;)